lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,234 @@
+    returned: success
+    type: dict
+    sample: {
+        'CompletionTime': '2018-01-24T00:47:33',
+        'Id': 'JID_XXXXXXXXXXXX',
+        'JobState': 'Completed',
+        'JobType': 'ImportConfiguration',
+        'Message': 'Successfully imported and applied Server Configuration Profile.',
+        'MessageId': 'SYSXXX',
+        'Name': 'Import Configuration',
+        'StartTime': 'TIME_NOW',
+        'Status': 'Success',
+    }
+"""
+
+
+from ansible.module_utils.remote_management.dellemc.dellemc_idrac import iDRACConnection
+from ansible.module_utils.basic import AnsibleModule
+try:
+    from omsdk.sdkcreds import UserCredentials
+    from omsdk.sdkfile import FileOnShare
+    HAS_OMSDK = True
+except ImportError:
+    HAS_OMSDK = False
+
+
+def _validate_catalog_file(params):
+    invalid, message = False, ""
+    if not params.get("catalog_file_name"):
+        invalid, message = True, "Invalid file name or invalid file extensions."
+    elif params.get("catalog_file_name"):
+        file_name = params.get("catalog_file_name").lower()
+        if not file_name.endswith("xml"):
+            invalid, message = True, "Invalid file name or invalid file extensions."
+    return invalid, message
+
+
+def run_update_fw_from_nw_share(idrac, module):
+    """
+    Update firmware from a network share
+    Keyword arguments:
+    idrac  -- iDRAC handle
+    module -- Ansible module
+    """
+
+    msg = {}
+    msg['changed'] = False
+    msg['msg'] = {}
+    err = False
+
+    try:
+        invalid, message = _validate_catalog_file(module.params)
+        if invalid:
+            err = True
+            msg['msg'] = message
+            return msg, err
+
+        upd_share = FileOnShare(remote=module.params['share_name'] + "/" + module.params['catalog_file_name'],
+                                mount_point=module.params['share_mnt'],
+                                isFolder=False,
+                                creds=UserCredentials(
+                                    module.params['share_user'],
+                                    module.params['share_pwd'])
+                                )
+
+        idrac.use_redfish = True
+        if '12' in idrac.ServerGeneration or '13' in idrac.ServerGeneration:
+            idrac.use_redfish = False
+
+        apply_update = True
+        msg['msg'] = idrac.update_mgr.update_from_repo(upd_share,
+                                                       apply_update,
+                                                       module.params['reboot'],
+                                                       module.params['job_wait'])
+
+        if "Status" in msg['msg']:
+            if msg['msg']['Status'] == "Success":
+                if module.params['job_wait'] is True:
+                    msg['changed'] = True
+            else:
+                err = True
+
+    except Exception as e:
+        err = True
+        msg['msg'] = "Error: %s" % str(e)
+
+    return msg, err
```

Comment:
This is a horrible API. Using return values to indicate errors is used in some ancient and low-level languages like C or Go which don't have nice built-in mechanisms for communicating errors to the caller.

In Python we raise exceptions for that and try/except at proper locations up the call stack.

Reply:
updated with `fail_json `